File: C++/283MoveZeros.cpp
#include <iostream>
#include <vector>
using namespace std;

class Solution {
public:
    void moveZeroes(vector<int>& nums) {
        int pos = 0;
        for(int i = 0; i < nums.size(); i++) {
            if(nums[i] != 0) {
                nums[pos++] = nums[i];
            }
        }
        while(pos < nums.size()) {
            nums[pos++] = 0;
        }
    }
};

void printVector(vector<int>& nums) {
    for(int i = 0; i< nums.size(); i++) {
        std::cout << nums[i] << " ";
    }
}

int main() {
    Solution sol;
    std::vector<int> nums = {0,1,0,3,12};
    sol.moveZeroes(nums);
    printVector(nums);
}

File: Swift/283MoveZeros.swift
class Solution {
    func moveZeroes(_ nums: inout [Int]) {
        var l = 0, r = 1
        if nums.count > 1 {
            // nums has 2 values or more so continue
            while r < nums.count {
                if nums[l] == 0 && nums[r] != 0 {
                    nums.swapAt(l, r)
                    l += 1
                    r += 1
                } else if nums[l] == 0 && nums[r] == 0 {
                    r += 1
                } else {
                    l += 1
                    r += 1
                }
            }
        } 
    }
}
